### ISSUE TITLE
blogdown addins do not work properly

### ISSUE THREAD

**Issue by michielvandijk** [rstudio/blogdown#201]
I am experiencing a problem with blogdown addins. It seems to work fine when a markdown related document is opened in Rstudio but otherwise not. When I click the "Update metadata" addin with nothing open in Rstudio I get the following error:

blogdown:::update_meta_addin()
Error in `Encoding<-`(`*tmp*`, value = "UTF-8") : 
  a character vector argument expected

When I then open index.Rmd (or any other markdown extension file) in Rstudio the addin opens and the following line is printed:

blogdown:::update_meta_addin()
Listening on http://127.0.0.1:3340

I am using the development version of blogdown.

> devtools::session_info("blogdown")
Session info --------------------------------------------------------------------------------
 setting  value                       
 version  R version 3.4.1 (2017-06-30)
 system   x86_64, mingw32             
 ui       RStudio (1.0.153)           
 language (EN)                        
 collate  English_United Kingdom.1252 
 tz       Europe/Berlin               
 date     2017-09-22                  

Packages ------------------------------------------------------------------------------------
 package   * version date       source                           
 backports   1.1.0   2017-05-22 CRAN (R 3.4.1)                   
 base64enc   0.1-3   2015-07-28 CRAN (R 3.4.1)                   
 bitops      1.0-6   2013-08-17 CRAN (R 3.4.1)                   
 blogdown    0.1.4   2017-09-21 Github (rstudio/blogdown@8c9c480)
 bookdown    0.5     2017-08-20 CRAN (R 3.4.1)                   
 caTools     1.17.1  2014-09-10 CRAN (R 3.4.1)                   
 digest      0.6.12  2017-01-27 CRAN (R 3.4.1)                   
 evaluate    0.10.1  2017-06-24 CRAN (R 3.4.1)                   
 graphics  * 3.4.1   2017-06-30 local                            
 grDevices * 3.4.1   2017-06-30 local                            
 highr       0.6     2016-05-09 CRAN (R 3.4.1)                   
 htmltools   0.3.6   2017-04-28 CRAN (R 3.4.1)                   
 httpuv      1.3.5   2017-07-04 CRAN (R 3.4.1)                   
 jsonlite    1.5     2017-06-01 CRAN (R 3.4.1)                   
 knitr       1.17    2017-08-10 CRAN (R 3.4.1)                   
 magrittr    1.5     2014-11-22 CRAN (R 3.4.1)                   
 markdown    0.8     2017-04-20 CRAN (R 3.4.1)                   
 methods   * 3.4.1   2017-06-30 local                            
 mime        0.5     2016-07-07 CRAN (R 3.4.1)                   
 Rcpp        0.12.12 2017-07-15 CRAN (R 3.4.1)                   
 rmarkdown   1.6     2017-06-15 CRAN (R 3.4.1)                   
 rprojroot   1.2     2017-01-16 CRAN (R 3.4.1)                   
 servr       0.7     2017-08-17 CRAN (R 3.4.1)                   
 stats     * 3.4.1   2017-06-30 local                            
 stringi     1.1.5   2017-04-07 CRAN (R 3.4.1)                   
 stringr     1.2.0   2017-02-18 CRAN (R 3.4.1)                   
 tools       3.4.1   2017-06-30 local                            
 utils     * 3.4.1   2017-06-30 local                            
 yaml        2.1.14  2016-11-12 CRAN (R 3.4.1)

**Comment by yihui** [comment]
@michielvandijk When you don't have any R Markdown or Markdown documents opened in RStudio, what do you expect the "Update Metadata" addin to do? In other words, update whose metadata? "Update" indicates you must have something existing to work on.

@jjallaire Anyway, I think the error message from **rstudioapi** is a little misleading when the source editor is empty:

```
> rstudioapi::getSourceEditorContext()
Error in `Encoding<-`(`*tmp*`, value = "UTF-8") : 
  a character vector argument expected
```

It probably should return `NULL` in this case.

**Comment by jjallaire** [comment]
@kevinushey

**Comment by michielvandijk** [comment]
That makes sense. I was a bit like a little boy pressing all the buttons of my new toy without actually knowing what I was doing. But indeed, the message is a bit misleading, in particular for first time users, as it looks like something is wrong, while in fact nothing should happen.

**Comment by yihui** [comment]
Excellent. Thanks, Kevin!

**Comment by AshesITR** [comment]
The Commit won't fix this. The new error will be

Error: 'x' should be a list of {range, text} pairs

The reason is that callFun("getSourceEditorContext") returns NULL, so that should be handled.

**Comment by yihui** [comment]
@AshesITR Could you tell us how you ran into this error? When your source editor is empty, you should not expect `rstudioapi::getSourceEditorContext()` to work.

**Comment by AshesITR** [comment]
@yihui I have some private workflow packages wich try to sensibly detect the current context package (akin to rprojroot but different). There is a function to create a new Rmd-script in the current context from a template. If this is the first such Rmd-script (or I simply closed all others), it errors out instead of falling back to getwd(). 

Returning NULL is much cleaner than erroring out IMHO since there is no way to check programmatically if the source editor is open (other than is.null(callFun("getSourceEditorContext"))). 
It's got nothing to do with blogdown though - google just turned up this thread and I noticed @kevinushey's commit to rstudioapi in an attempt to fix the problem.
